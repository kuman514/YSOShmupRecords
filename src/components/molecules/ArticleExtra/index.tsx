import React from 'react';
import styled from 'styled-components';

import { ShmupRecord } from '^/types';

const Root = styled.div`
  width: 100%;
`;

const Column = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  align-items: center;
  column-gap: 15px;
`;

const ColumnTitle = styled.div`
  width: 180px;
  font-weight: 600;
  font-size: 16px;
`;

const ColumnContent = styled.div`
  font-size: 16px;
  font-weight: 400;
`;

interface Props {
  comment: ShmupRecord['comment'];
  youtubeUrl?: ShmupRecord['youtubeUrl'];
}

export function ArticleExtra({ comment, youtubeUrl }: Props) {
  /**
   * @todo
   * Replace 유튜브 영상 칸 into the real youtube embed
   */
  const renderYoutube =
    youtubeUrl !== undefined ? (
      <Column>
        <ColumnTitle>유튜브 링크</ColumnTitle>
        <ColumnContent>유튜브 영상 칸</ColumnContent>
      </Column>
    ) : null;

  return (
    <Root>
      <Column>
        <ColumnTitle>코멘터리</ColumnTitle>
        <ColumnContent>{comment}</ColumnContent>
      </Column>
      {renderYoutube}
    </Root>
  );
}
