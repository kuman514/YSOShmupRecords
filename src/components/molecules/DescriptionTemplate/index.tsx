import styled from 'styled-components';

import { DescriptionListItem } from '^/types';

const Root = styled.div`
  display: flex;
  flex-direction: column;
  row-gap: 16px;
  padding-left: 15px;
  padding-right: 15px;
`;

const Title = styled.h1`
  font-size: 36px;
  font-weight: 700;
`;

const ListItem = styled.li`
  font-size: 16px;
  font-weight: 400;

  &:not(:last-of-type) {
    margin-bottom: 8px;
  }
`;

interface Props {
  title: string;
  descriptionListItems: DescriptionListItem[];
}

export function DescriptionTemplate({ title, descriptionListItems }: Props) {
  const renderDescriptionListItems = descriptionListItems.map(
    ({ id, description }) => <ListItem key={id}>{description}</ListItem>
  );

  return (
    <Root>
      <Title>{title}</Title>
      <ul>{renderDescriptionListItems}</ul>
    </Root>
  );
}
