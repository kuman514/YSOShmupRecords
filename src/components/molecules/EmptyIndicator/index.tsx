import { useNavigate } from 'react-router-dom';
import styled from 'styled-components';

import EmptyPngUrl from '^/assets/images/empty.png';
import { Button } from '^/components/atoms/Button';
import { ButtonType } from '^/types';

const Root = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
  row-gap: 16px;
`;

const Image = styled.img`
  width: 100%;
  max-width: 400px;
`;

const EmptyInformationTitle = styled.h2`
  font-size: 36px;
  font-weight: 700;
`;

export function EmptyIndicator() {
  const navigate = useNavigate();

  return (
    <Root>
      <Image src={EmptyPngUrl} alt="에러 발생" />
      <EmptyInformationTitle>
        현재 등록된 기록이 없습니다.
      </EmptyInformationTitle>
      <Button
        type={ButtonType.SOLID}
        isDisabled={false}
        onClick={() => {
          navigate('/');
        }}
        customStyle={{
          width: 'fit-content',
        }}
      >
        타이틀 화면으로 돌아가기
      </Button>
      <Button
        type={ButtonType.SOLID}
        isDisabled={false}
        onClick={() => {
          window.open('https://open.kakao.com/me/YSOShmupRecords', '_blank');
        }}
        customStyle={{
          width: 'fit-content',
        }}
      >
        소통창구 (카카오 오픈채팅방)
      </Button>
    </Root>
  );
}
