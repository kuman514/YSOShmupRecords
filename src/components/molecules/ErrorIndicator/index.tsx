import React from 'react';
import styled from 'styled-components';
import { useNavigate } from 'react-router-dom';

import ErrorPngUrl from '^/assets/images/error.png';
import { Button } from '^/components/atoms/Button';
import { ButtonType } from '^/types';

const Root = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
  align-items: center;
  text-align: center;
  row-gap: 16px;
`;

const Image = styled.img`
  width: 100%;
  max-width: 400px;
`;

const ErrorOrEmptyInformationTitle = styled.h2`
  font-size: 36px;
  font-weight: 700;
`;

const ErrorOrEmptyInformation = styled.span`
  font-size: 16px;
  font-weight: 400;
`;

interface Props {
  title: string;
}

export function ErrorIndicator({ title }: Props) {
  const navigate = useNavigate();

  return (
    <Root>
      <Image src={ErrorPngUrl} alt="에러 발생" />
      <ErrorOrEmptyInformationTitle>{title}</ErrorOrEmptyInformationTitle>
      <ErrorOrEmptyInformation>
        인터넷 연결 상태와 현재 주소를 확인해주세요.
      </ErrorOrEmptyInformation>
      <ErrorOrEmptyInformation>
        서버 상의 에러일 수 있으니 잠시 후 다시 시도해주세요.
      </ErrorOrEmptyInformation>
      <ErrorOrEmptyInformation>
        문제 발생 시 아래 카카오 오픈채팅방 버튼으로 문의해주세요.
      </ErrorOrEmptyInformation>
      <Button
        type={ButtonType.SOLID}
        isDisabled={false}
        onClick={() => {
          navigate('/');
        }}
        customStyle={{
          width: 'fit-content',
        }}
      >
        타이틀 화면으로 돌아가기
      </Button>
      <Button
        type={ButtonType.SOLID}
        isDisabled={false}
        onClick={() => {
          window.open('https://open.kakao.com/me/YSOShmupRecords', '_blank');
        }}
        customStyle={{
          width: 'fit-content',
        }}
      >
        소통창구 (카카오 오픈채팅방)
      </Button>
    </Root>
  );
}
