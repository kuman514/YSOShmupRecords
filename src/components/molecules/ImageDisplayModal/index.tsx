import React from 'react';

import { ImageDisplay } from '^/components/atoms/ImageDisplay';
import { Overlay } from '^/components/atoms/Overlay';

interface Props {
  imageUrl: string;
  onExit(): void;
}

export function ImageDisplayModal({ imageUrl, onExit }: Props) {
  return (
    <Overlay onClick={onExit}>
      <ImageDisplay imageUrl={imageUrl} onClickClose={onExit} />
    </Overlay>
  );
}
