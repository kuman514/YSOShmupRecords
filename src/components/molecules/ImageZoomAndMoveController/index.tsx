import React, { useState } from 'react';
import styled from 'styled-components';

const Root = styled.div`
  width: 100%;
  height: 100%;
  cursor: grab;
`;

const Image = styled.img`
  width: 100%;
  height: 100%;
  object-fit: contain;

  -webkit-user-drag: none;
  -khtml-user-drag: none;
  -moz-user-drag: none;
  -o-user-drag: none;
  user-drag: none;

  -webkit-user-select: none;
  -khtml-user-select: none;
  -moz-user-select: none;
  -o-user-select: none;
  user-select: none;
`;

interface Props {
  imageUrl: string;
}

export function ImageZoomAndMoveController({ imageUrl }: Props) {
  const [recentTouchDist, setRecentTouchDist] = useState<number>(-1);
  const [position, setPosition] = useState<number[]>([0, 0]);
  const [scale, setScale] = useState<number>(1);

  function moveImage(dx: number, dy: number) {
    setPosition([
      position[0] + dx * (1 / scale),
      position[1] + dy * (1 / scale),
    ]);
  }

  return (
    <Root
      onClick={(event) => {
        event.stopPropagation();
        event.nativeEvent.stopPropagation();
        event.nativeEvent.stopImmediatePropagation();
      }}
      onWheel={(event) => {
        const newScale = scale - event.deltaY * 0.001;
        const finalScale = Math.min(Math.max(0.3, newScale), 10);
        setScale(finalScale);
      }}
      onMouseMove={(event) => {
        if (event.buttons !== 1) {
          return;
        }
        moveImage(event.movementX, event.movementY);
      }}
      onTouchStart={(touchStartEvent) => {
        function handleOnTouchMove(touchMoveEvent: TouchEvent) {
          if (touchMoveEvent.cancelable) {
            touchMoveEvent.preventDefault();
          }

          const deltaX =
            touchMoveEvent.touches[0].pageX - touchStartEvent.touches[0].pageX;
          const deltaY =
            touchMoveEvent.touches[0].pageY - touchStartEvent.touches[0].pageY;
          moveImage(deltaX, deltaY);

          if (touchMoveEvent.touches.length === 2) {
            const firstPageX = touchMoveEvent.touches[0].pageX;
            const firstPageY = touchMoveEvent.touches[0].pageY;
            const secondPageX = touchMoveEvent.touches[1].pageX;
            const secondPageY = touchMoveEvent.touches[1].pageY;

            const left = Math.min(firstPageX, secondPageX);
            const right = Math.max(firstPageX, secondPageX);
            const top = Math.min(firstPageY, secondPageY);
            const bottom = Math.max(firstPageY, secondPageY);

            const horizontalDist = right - left;
            const verticalDist = bottom - top;

            const dist = Math.sqrt(
              horizontalDist * horizontalDist + verticalDist * verticalDist
            );

            if (recentTouchDist > 0) {
              const newScale = scale - (recentTouchDist - dist) / 100;
              const finalScale = Math.min(Math.max(0.3, newScale), 10);
              setScale(finalScale);
            }
          } else {
            setRecentTouchDist(-1);
          }
        }

        function handleOnTouchEnd() {
          document.removeEventListener('touchmove', handleOnTouchMove);
        }

        document.addEventListener('touchmove', handleOnTouchMove);
        document.addEventListener('touchend', handleOnTouchEnd, { once: true });
      }}
    >
      <Image
        src={imageUrl}
        alt="Record image in expanded contain size"
        style={{
          transform: `scale(${scale}) translate(${position[0]}px, ${position[1]}px)`,
        }}
      />
    </Root>
  );
}
