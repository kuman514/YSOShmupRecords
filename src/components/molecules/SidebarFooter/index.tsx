import React from 'react';
import styled from 'styled-components';

import { Button } from '^/components/atoms/Button';
import { ButtonType } from '^/types';

const Root = styled.div`
  display: flex;
  flex-direction: row;
  flex-wrap: wrap;
  justify-content: center;
  align-items: center;

  row-gap: 8px;
  column-gap: 20px;
  padding: 16px 16px 16px 32px;

  color: var(--white-color);

  font-size: 16px;
  font-weight: 300;
`;

const ContactButtonArray = styled.div`
  display: flex;
  flex-direction: row;
`;

const ContactButtonWrapper = styled.div`
  padding: 0 10px;

  &:not(:last-of-type) {
    border-right: 1px solid var(--white-color);
  }
`;

export function SidebarFooter() {
  return (
    <Root>
      <div>YSO as kuman514</div>
      <ContactButtonArray>
        <ContactButtonWrapper>
          <Button
            type={ButtonType.CLEAR}
            isDisabled={false}
            onClick={() => {
              window.open(
                'https://open.kakao.com/me/YSOShmupRecords',
                '_blank'
              );
            }}
          >
            소통창구 (카카오 오픈채팅방)
          </Button>
        </ContactButtonWrapper>
      </ContactButtonArray>
    </Root>
  );
}
