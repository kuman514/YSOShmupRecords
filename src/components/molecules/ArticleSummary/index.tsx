import React, { useState } from 'react';
import styled from 'styled-components';
import { useToast } from '@chakra-ui/react';

import { ButtonType, ShmupRecord } from '^/types';
import { Thumbnail } from '^/components/atoms/Thumbnail';
import { convertDateToString } from '^/utils/date-to-string';
import { textsForArticle } from '^/constants/texts';
import { ImageDisplayModal } from '^/components/molecules/ImageDisplayModal';
import { NavRouteTitle } from '^/components/atoms/NavRouteTitle';
import { Button } from '^/components/atoms/Button';
import { ReactComponent as RawLinkSvg } from '^/assets/icons/link.svg';
import { ReactComponent as RawTwitterSvg } from '^/assets/icons/twitter.svg';

const Root = styled.div`
  width: 100%;
  display: flex;
  flex-direction: column;
  flex-wrap: wrap;
  row-gap: 16px;
`;

const SummaryArea = styled.div`
  display: flex;
  flex-direction: row;
  align-items: center;
  flex-wrap: wrap;
  gap: 16px;
`;

const Title = styled.h1`
  font-size: 36px;
  font-weight: 700;
`;

const SummaryDescription = styled.div`
  display: flex;
  flex-direction: column;
  gap: 8px;
`;

const List = styled.ul`
  padding-left: 15px;

  & > li:not(:last-of-type) {
    margin-bottom: 8px;
  }
`;

const ListItemTitle = styled.span`
  font-size: 16px;
  font-weight: 600;
`;

const ListItemContent = styled.span`
  font-size: 16px;
  font-weight: 400;

  &::before {
    content: ' ';
    white-space: pre;
  }
`;

const SpecialTag = styled.span`
  & + &::before {
    content: ', ';
    white-space: pre;
  }
`;

const ShareButtonList = styled.div`
  display: flex;
  flex-direction: row;
  gap: 8px;
`;

const LinkSvg = styled(RawLinkSvg)`
  width: 24px;
  height: 24px;
`;

const TwitterSvg = styled(RawTwitterSvg)`
  width: 24px;
  height: 24px;
`;

interface Props {
  record: ShmupRecord;
}

export function ArticleSummary({ record }: Props) {
  const toast = useToast();
  const [isImageModalShow, setIsImageModalShow] = useState<boolean>(false);

  const renderSpecialTags =
    record.specialTags !== undefined && record.specialTags.length > 0 ? (
      <li>
        <ListItemTitle>특이사항</ListItemTitle>
        <ListItemContent>
          {record.specialTags.map((specialTag) => (
            <SpecialTag key={specialTag}>{specialTag}</SpecialTag>
          ))}
        </ListItemContent>
      </li>
    ) : null;

  const renderList = (
    <List>
      <li>
        <ListItemTitle>최종 스테이지</ListItemTitle>
        <ListItemContent>{record.stage}</ListItemContent>
      </li>
      <li>
        <ListItemTitle>최종 점수</ListItemTitle>
        <ListItemContent>{record.score}점</ListItemContent>
      </li>
      <li>
        <ListItemTitle>수단/장소</ListItemTitle>
        <ListItemContent>
          {textsForArticle[record.byWhat] ?? record.byWhat}
        </ListItemContent>
      </li>
      {renderSpecialTags}
    </List>
  );

  const renderImageModal = isImageModalShow ? (
    <ImageDisplayModal
      imageUrl={record.originalImageUrl}
      onExit={() => {
        setIsImageModalShow(false);
      }}
    />
  ) : null;

  return (
    <Root>
      <Title>{convertDateToString(record.when)}</Title>
      <NavRouteTitle />
      <SummaryArea>
        <Thumbnail
          imageSrc={record.thumbnailUrl}
          altText={`${record.subjectId} ${convertDateToString(record.when)} ${
            record.stage
          }스테이지 ${record.score}점`}
          onClick={() => {
            setIsImageModalShow(true);
          }}
        />
        <SummaryDescription>
          {renderList}
          <ShareButtonList>
            <Button
              type={ButtonType.ROUND_LINE}
              isDisabled={false}
              onClick={() => {
                navigator.clipboard.writeText(window.location.href);
                toast({
                  duration: 2000,
                  title: '링크 복사 완료.',
                  description: '공유하기 원하시는 곳에 붙여넣으십시오.',
                  status: 'success',
                  isClosable: true,
                });
              }}
              customStyle={{
                position: 'relative',
                display: 'flex',
                padding: '10px',
                borderRadius: '50%',
              }}
            >
              <LinkSvg />
            </Button>
            <Button
              type={ButtonType.ROUND_LINE}
              isDisabled={false}
              onClick={() => {
                const urlToUri = encodeURI(window.location.href);
                const textToUri = encodeURI(
                  `${convertDateToString(record.when)}, ${
                    textsForArticle[record.byWhat]
                  }에서 플레이한 ${textsForArticle[record.subjectId]}에서, ${
                    record.score
                  }점으로 ${record.stage} 달성!`
                );

                const tweet = `https://twitter.com/intent/tweet?url=${urlToUri}&text=${textToUri}`;
                window.open(tweet, '_blank');
              }}
              customStyle={{
                display: 'flex',
                padding: '10px',
                borderRadius: '50%',
              }}
            >
              <TwitterSvg />
            </Button>
          </ShareButtonList>
        </SummaryDescription>
      </SummaryArea>
      {renderImageModal}
    </Root>
  );
}
