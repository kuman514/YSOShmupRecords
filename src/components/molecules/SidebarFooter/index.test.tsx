import { cleanup, render } from '@testing-library/react';
import 'jest-styled-components';
import { beforeEach, describe, expect, it } from 'vitest';

import { SidebarFooter } from '^/components/molecules/SidebarFooter';

describe('SidebarFooter', () => {
  beforeEach(() => {
    cleanup();
  });

  it('should have a snapshot match for its appearance', () => {
    const { container } = render(<SidebarFooter />);
    expect(container.firstChild).toMatchSnapshot();
  });
});
