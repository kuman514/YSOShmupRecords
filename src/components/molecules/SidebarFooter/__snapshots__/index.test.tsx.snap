// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SidebarFooter > should have a snapshot match for its appearance 1`] = `
.c3 {
  all: unset;
  box-sizing: border-box;
  font-family: Inter;
  font-size: 16px;
  font-weight: 500;
  padding: 8px 16px;
  color: #ffffff;
  background-color: transparent;
  border: none;
  border-radius: 16px;
  cursor: pointer;
}

.c3:hover {
  background-color: var(--hovering-color);
  color: #ffffff;
}

.c3:disabled {
  border: none;
  color: var(--gray-color);
  background-color: transparent;
  text-decoration: line-through;
}

.c0 {
  display: flex;
  flex-direction: row;
  justify-content: center;
  align-items: center;
  column-gap: 20px;
  padding: 16px 16px 16px 32px;
  color: #ffffff;
  font-size: 16px;
  font-weight: 300;
}

.c1 {
  display: flex;
  flex-direction: row;
}

.c2 {
  padding: 0 10px;
}

.c2:not(:last-of-type) {
  border-right: 1px solid #ffffff;
}

<div
  class="c0"
>
  <div>
    YSO as kuman514
  </div>
  <div
    class="c1"
  >
    <div
      class="c2"
    >
      <button
        class="c3"
      >
        소통창구 (카카오 오픈채팅방)
      </button>
    </div>
  </div>
</div>
`;
